src/app_callback_service.rs

```diff
@@ -1,9 +1,30 @@
+use json::JsonValue;
+use mongodb::Collection;
 use tonic::{Request, Response, Status};
 
+use bson::Uuid;
 use dapr::{appcallback::*, dapr::dapr::proto::runtime::v1::app_callback_server::AppCallback};
 
-#[derive(Default)]
-pub struct AppCallbackService {}
+use crate::foreign_types::ProductVariant;
+
+pub struct AppCallbackService {
+    pub collection: Collection<ProductVariant>,
```

Comment:
While I am admittedly not good at Rust, I wonder why you save the known product variants in-memory?
At least, that's what it looks like to me.
If that's the case, perhaps don't, just store it in the db and ask the db if it exists.
Otherwise we get problems with restarting the service.
Anyway, what happens when a product variant is created while this service is offline?

Reply:
as it is a `use mongodb::Collection;`, probably it is just to access the db?

Reply:
Yes, this is just the db connection.

Reply:
The library enables the MongoDB Collections to be typed.